src/sage/stats/distributions/discrete_gaussian_lattice.py

```diff
@@ -222,51 +228,217 @@ def _normalisation_factor_zz(self, tau=3):
+                return R(exp(-pi**2 * (2 * sigma**2) * x))
+
+            return R(exp(-x / (2 * sigma**2)))
+
+        if not self.is_spherical:
+            # TODO: This is only a poor approximation placeholder.
+            # It should be easy to implement, since the Fourier transform
+            # is essentially the same, but I can't figure out how to
+            # tweak the `.qfrep` call below correctly.
+            from warnings import warn
+            warn("Note: `_normalisation_factor_zz` has not been properly "
+                 "implemented for non-spherical distributions.")
+            import itertools
+            from sage.functions.log import log
+            basis = self.B.LLL()
+            base = vector(ZZ, [v.round() for v in basis.solve_left(self._c)])
+            BOUND = max(1, (self._RR(log(10**4, self.n)).ceil() - 1) // 2)
+            if BOUND > 10:
+                BOUND = 10
+            coords = itertools.product(range(-BOUND, BOUND + 1), repeat=self.n)
+            return sum(self.f((vector(u) + base) * self.B) for u in coords)
+
+        if self.B.nrows() != self.B.ncols():
+            raise NotImplementedError("basis must be a square matrix for now")
+
+        if self.is_spherical and not self._c_in_lattice:
+            raise NotImplementedError("lattice must contain 0 for now")
+
+        if self.B.base_ring() != ZZ:
+            raise NotImplementedError("lattice must be integral for now")
 
-        f = self.f
-        n = self.B.ncols()
         sigma = self._sigma
-        return sum(f(x) for x in _iter_vectors(n, -ceil(tau * sigma),
-                                               ceil(tau * sigma)))
+        prec = DiscreteGaussianDistributionLatticeSampler.compute_precision(
+            prec, sigma
+        )
+        R = RealField(prec=prec)
+        if sigma > 1:
+            det = self.B.det()
+            norm_factor = (sigma * sqrt(2 * pi))**self.n / det
+        else:
+            det = 1
+            norm_factor = 1
+
+        # qfrep computes theta series of a quadratic form, which is *half* the
+        # generating function of number of vectors with given norm (and no 0)
+        Q = self.Q
+        if tau is not None:
+            freq = Q.__pari__().qfrep(tau * sigma, 0)
+            res = R(1)
+            for x, fq in enumerate(freq):
+                res += 2 * ZZ(fq) * f_or_hat((x + 1) / det**self.n)
+            return R(norm_factor * res)
+
+        res = R(1)
+        bound = 0
+        # There might still be precision issue but whatever
+        while True:
+            bound += 1
+            cnt = ZZ(Q.__pari__().qfrep(bound, 0)[bound - 1])
+            inc = 2 * cnt * f_or_hat(bound / det**self.n)
+            if cnt > 0 and res == res + inc:
+                return R(norm_factor * res)
+            res += inc
+
+    @cached_method
+    def _maximal_r(self):
+        r"""
+        This function computes the largest value `r > 0` such that `\Sigma - r^2BB^@`
```

Comment:
Is the `@` here intentional or should it be a `T`?

Reply:
It's a `T`, thanks.